### DISCUSSION TITLE
How to use Oauth with client

### DISCUSSION THREAD

**Discussion by mzawalski-vilisoft** [cloudbees/zendesk-java-client#520] [Q&A] [answered]
Could anyone tell me how to build and reference `zendesk-java-client` client to auth via Oauth client on behalf of end-user using Request Api as described in [Zendesk Docs](https://developer.zendesk.com/documentation/ticketing/using-the-zendesk-api/making-api-requests-on-behalf-of-end-users/)?

**Comment by PierreBtz** [comment]
Hey @mzawalski-vilisoft,  it looks like what you need is to set the `X-On-Behalf-Of` header. You could do so by passing your own `AsyncHttpClient` setup with this header.

Here is a quick example, note that I couldn't test it since I don't have access to an instance setup for impersonation:

```
        var zdUrl = "";
        // TODO your code to get the impersonation token, can be done with a simple call with any http client
        var impersonationToken = "";

        var cilentConfig = new DefaultAsyncHttpClientConfig.Builder()
            .setFollowRedirect(true)
            // TODO untested, reading the code of the async client it seems like this way of adding a header should work
            // but there might be other (better?) ways
            .addRequestFilter(new RequestFilter() {
                @Override
                public <T> FilterContext<T> filter(FilterContext<T> ctx) {
                    ctx.getRequest().getHeaders().add("X-On-Behalf-Of", "TODO");
                    return ctx;
                }
            })
            .build();
        var impersonatingClient = new Zendesk.Builder(zdUrl)
            .setOauthToken(impersonationToken)
            .setClient(new DefaultAsyncHttpClient(cilentConfig))
            .build();
```

**Comment by mzawalski-vilisoft** [comment, marked as answer]
Hi @PierreBtz, thank you for the detailed answer with the example. The problem for me was how to generate such a token 😅 
Step by step for others:

1. Create an OAuth client via `Zendesk -> Admin Center -> Settings ->Apps and Integrations -> Zendesk API -> OAuth clients -> Add`
2. Next list the clients via Rest API: `GET /api/v2/oauth/clients` and copy the **id** of you new client
3. Generate token with scopes for the client via Rest API: `POST /api/v2/oauth/tokens` with body:
```
{
    "token": {
        "client_id": "COPIED_ID",
        "scopes": ["read", "write", "impersonate"]
    }
}
```
Copy the **full_token** from the response
4. Create your Zendesk client:
```
        Zendesk zendesk = new Zendesk.Builder("https://XXXXXXXXXX.zendesk.com")
                .setOauthToken("FULL_TOKEN_FROM_PREVIOS_STEP")
                .addHeader("X-On-Behalf-Of", "END_USER_EMAIL_YOU_WANT_TO_IMPERSONATE")
                .build();
```
Hope it will help someone :)
